CIS cannot discover service endpoints when the targetPort is referred name instead of number

#### Setup Details
 CIS Version : 2.10.1     
 Build: f5networks/k8s-bigip-ctlr:2.10.1      
 BIGIP Version: BIG-IP 15.1.3 Build 0.0.11 Final     
 AS3 Version: f5-appsvcs-3.38.0-4.noarch   
 Agent Mode: AS3/CCCL       
 Orchestration: K8S/OSCP      
 Orchestration Version:  Kubernetes  v1.22.13     
 Pool Mode: Cluster/Nodeport      
 Additional Setup details: Rancher/Calico

#### Description  
When the service has name in TargetPort then the members in the load balancer is empty.


#### Steps To Reproduce

1) Deploy rabbitmq app using helm; chart version  rabbitmq-8.31.5 and app version  3.9.15. And look at the service created by the app.
```
kg svc/rabbitmq -o yaml -n data
apiVersion: v1
kind: Service
metadata:
  annotations:
    meta.helm.sh/release-name: rabbitmq
    meta.helm.sh/release-namespace: data
  creationTimestamp: "2022-04-19T18:52:58Z"
  labels:
    app.kubernetes.io/instance: rabbitmq
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: rabbitmq
    helm.sh/chart: rabbitmq-8.31.5
  name: rabbitmq
  namespace: data
  resourceVersion: "240109529"
  uid: 55d1b599-1895-460d-b403-0c7eb7cc66f0
spec:
  clusterIP: 10.43.198.109
  clusterIPs:
  - 10.43.198.109
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: amqp
    port: 5672
    protocol: TCP
    targetPort: amqp
  - name: amqp-ssl
    port: 5671
    protocol: TCP
    targetPort: amqp-ssl
  - name: epmd
    port: 4369
    protocol: TCP
    targetPort: epmd
  - name: dist
    port: 25672
    protocol: TCP
    targetPort: dist
  - name: http-stats
    port: 15672
    protocol: TCP
    targetPort: stats
  selector:
    app.kubernetes.io/instance: rabbitmq
    app.kubernetes.io/name: rabbitmq
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```
2) Create the following vip
```
apiVersion: cis.f5.com/v1
kind: VirtualServer
metadata:
  creationTimestamp: "2023-01-17T16:19:47Z"
  generation: 1
  labels:
    f5cr: "true"
  name: dev1.messagebus.p2.internal
  namespace: data
  resourceVersion: "369930540"
  uid: 5927e4c0-73b6-4c6c-9491-1fdbfa48b319
spec:
  pools:
  - monitor:
      interval: 20
      recv: ""
      send: /
      timeout: 31
      type: https
    path: /
    service: rabbitmq
    servicePort: 15672
  tlsProfileName: dev1.messagebus.p2.internal
  virtualServerAddress: 10.40.14.207
  virtualServerName: dev1_messagebus_p2_internal_vs
status:
  status: Ok
  vsAddress: None
```
3) Check the logs in the f5 pod

```
2023/01/17 15:00:41 [DEBUG] Processing Key: &{data TLSProfile dev1.messagebus.p2.internal 0xc0005df500 Create}
2023/01/17 15:00:41 [INFO] Enqueueing TLSProfile: &{{ } {dev1.messagebus.p2.internal  data  4d6c3de4-a72a-486b-a5af-bd34c8665d23 369880900 1 2023-01-17 15:00:41 +0000 UTC <nil> <nil> map[f5cr:true] map[] [] []  [{kubectl-create Update cis.f5.com/v1 2023-01-17 15:00:41 +0000 UTC FieldsV1 {"f:metadata":{"f:labels":{".":{},"f:f5cr":{}}},"f:spec":{".":{},"f:tls":{".":{},"f:clientSSL":{},"f:reference":{},"f:serverSSL":{},"f:termination":{}}}}}]} {[] {reencrypt /Common/dev1.messagebus.p2.internal_ssl_client /Common/dev1.messagebus.p2.internal_ssl_server bigip}}}
2023/01/17 15:00:41 [INFO] Change in TLSProfile dev1.messagebus.p2.internal does not effect any VirtualServer
2023/01/17 15:00:42 [DEBUG] Enqueueing VirtualServer: &{{ } {dev1.messagebus.p2.internal  data  5e503746-f5c8-4444-b7d5-60cd3b2b3714 369880910 1 2023-01-17 15:00:42 +0000 UTC <nil> <nil> map[f5cr:true] map[] [] []  [{kubectl-create Update cis.f5.com/v1 2023-01-17 15:00:42 +0000 UTC FieldsV1 {"f:metadata":{"f:labels":{".":{},"f:f5cr":{}}},"f:spec":{".":{},"f:pools":{},"f:tlsProfileName":{},"f:virtualServerAddress":{},"f:virtualServerName":{}}}}]} {  10.40.14.207  dev1_messagebus_p2_internal_vs 0 0 [{ / rabbitmq 15672  {https /  20 31 0  } []   }] dev1.messagebus.p2.internal     [] [] []      {{ }      []  } []} { }}
2023/01/17 15:00:42 [DEBUG] Processing Key: &{data VirtualServer dev1.messagebus.p2.internal 0xc001862000 Create}
2023/01/17 15:00:42 [DEBUG] Process all the Virtual Servers which share same VirtualServerAddress
2023/01/17 15:00:42 [DEBUG] Processing Virtual Server dev1.messagebus.p2.internal for port 443
2023/01/17 15:00:42 [DEBUG] Configured rule: {vs__rabbitmq_15672_data  0 [0xc0002812c0] []}
2023/01/17 15:00:42 [DEBUG] Configured policy: {dev1_messagebus_p2_internal_vs_443__policy data  [forwarding]  true [http] [0xc000281440] /Common/first-match}
2023/01/17 15:00:42 [DEBUG] Processing  BIGIP referenced profiles for 'VirtualServer' 'data'/'dev1.messagebus.p2.internal'
2023/01/17 15:00:42 [DEBUG] Updated BIGIP referenced profiles for 'VirtualServer' 'data'/'dev1.messagebus.p2.internal'
2023/01/17 15:00:42 [DEBUG] Updated Virtual dev1.messagebus.p2.internal with TLSProfile dev1.messagebus.p2.internal
2023/01/17 15:00:42 [ERROR] [CORE]Endpoints could not be fetched for service rabbitmq with targetPort 0
```



#### Expected Result 
The endpoints should be detected by f5


#### Actual Result  
Endpoints could not be fetched for service rabbitmq with targetPort 0

#### Diagnostic Information
```buildoutcfg

2023/01/17 15:01:06 [DEBUG] Enqueueing VirtualServer: &{{ } {dev1.messagebus.p2.internal  data  5e503746-f5c8-4444-b7d5-60cd3b2b3714 369881142 1 2023-01-17 15:00:42 +0000 UTC <nil> <nil> map[f5cr:true] map[] [] []  [{kubectl-create Update cis.f5.com/v1 2023-01-17 15:00:42 +0000 UTC FieldsV1 {"f:metadata":{"f:labels":{".":{},"f:f5cr":{}}},"f:spec":{".":{},"f:pools":{},"f:tlsProfileName":{},"f:virtualServerAddress":{},"f:virtualServerName":{}}}} {k8s-bigip-ctlr.real Update cis.f5.com/v1 2023-01-17 15:01:06 +0000 UTC FieldsV1 {"f:status":{".":{},"f:status":{},"f:vsAddress":{}}}}]} {  10.40.14.207  dev1_messagebus_p2_internal_vs 0 0 [{ / rabbitmq 15672  {https /  20 31 0  } []   }] dev1.messagebus.p2.internal     [] [] []      {{ }      []  } []} {None Ok}}
2023/01/17 15:01:06 [DEBUG] Processing Key: &{data VirtualServer dev1.messagebus.p2.internal 0xc001862700 Update}
2023/01/17 15:01:06 [DEBUG] Process all the Virtual Servers which share same VirtualServerAddress
2023/01/17 15:01:06 [DEBUG] Processing Virtual Server dev1.messagebus.p2.internal for port 443
2023/01/17 15:01:06 [DEBUG] Configured rule: {vs__rabbitmq_15672_data  0 [0xc001d8a540] []}
2023/01/17 15:01:06 [DEBUG] Configured policy: {dev1_messagebus_p2_internal_vs_443__policy data  [forwarding]  true [http] [0xc001d8a5a0] /Common/first-match}
2023/01/17 15:01:06 [DEBUG] Processing  BIGIP referenced profiles for 'VirtualServer' 'data'/'dev1.messagebus.p2.internal'
2023/01/17 15:01:06 [DEBUG] Updated BIGIP referenced profiles for 'VirtualServer' 'data'/'dev1.messagebus.p2.internal'
2023/01/17 15:01:06 [DEBUG] Updated Virtual dev1.messagebus.p2.internal with TLSProfile dev1.messagebus.p2.internal
2023/01/17 15:01:06 [ERROR] [CORE]Endpoints could not be fetched for service rabbitmq with targetPort 0
2023/01/17 15:01:06 [DEBUG] Finished syncing virtual servers &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:dev1.messagebus.p2.internal GenerateName: Namespace:data SelfLink: UID:5e503746-f5c8-4444-b7d5-60cd3b2b3714 ResourceVersion:369881142 Generation:1 CreationTimestamp:2023-01-17 15:00:42 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[f5cr:true] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName: ManagedFields:[{Manager:kubectl-create Operation:Update APIVersion:cis.f5.com/v1 Time:2023-01-17 15:00:42 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:labels":{".":{},"f:f5cr":{}}},"f:spec":{".":{},"f:pools":{},"f:tlsProfileName":{},"f:virtualServerAddress":{},"f:virtualServerName":{}}}} {Manager:k8s-bigip-ctlr.real Operation:Update APIVersion:cis.f5.com/v1 Time:2023-01-17 15:01:06 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:status":{".":{},"f:status":{},"f:vsAddress":{}}}}]} Spec:{Host: HostGroup: VirtualServerAddress:10.40.14.207 IPAMLabel: VirtualServerName:dev1_messagebus_p2_internal_vs VirtualServerHTTPPort:0 VirtualServerHTTPSPort:0 Pools:[{Name: Path:/ Service:rabbitmq ServicePort:15672 NodeMemberLabel: Monitor:{Type:https Send:/ Recv: Interval:20 Timeout:31 TargetPort:0 Name: Reference:} Monitors:[] Rewrite: Balance: ServiceNamespace:}] TLSProfileName:dev1.messagebus.p2.internal HTTPTraffic: SNAT: WAF: RewriteAppRoot: AllowVLANs:[] IRules:[] ServiceIPAddress:[] PolicyName: PersistenceProfile: ProfileMultiplex: DOS: BotDefense: Profiles:{TCP:{Client: Server:} UDP: HTTP: HTTP2: RewriteProfile: PersistenceProfile: LogProfiles:[] ProfileL4: ProfileMultiplex:} AllowSourceRange:[]} Status:{VSAddress:None StatusOk:Ok}} (200.862µs)


Note: Sanitize the data. For example, be mindful of IPs, ports, application names and URLs
Note: The following F5 article outlines the information required when opening an issue.
https://support.f5.com/csp/article/K60974137
```  

#### Observations (if any)

@trinaths I updated the svc and still I get the same error. 

Service
```
 kg svc/rabbitmq -o yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    meta.helm.sh/release-name: rabbitmq
    meta.helm.sh/release-namespace: data
  creationTimestamp: "2022-09-05T20:08:06Z"
  labels:
    app.kubernetes.io/instance: rabbitmq
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: rabbitmq
    helm.sh/chart: rabbitmq-8.31.5
  name: rabbitmq
  namespace: data
  resourceVersion: "220276321"
  uid: 0bc80adc-996b-4f98-962c-8798e5e1b2a9
spec:
  clusterIP: 10.43.124.161
  clusterIPs:
  - 10.43.124.161
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: amqp
    port: 5672
    protocol: TCP
    targetPort: amqp
  - name: amqp-ssl
    port: 5671
    protocol: TCP
    targetPort: amqp-ssl
  - name: epmd
    port: 4369
    protocol: TCP
    targetPort: epmd
  - name: dist
    port: 25672
    protocol: TCP
    targetPort: dist
  - name: http-stats
    port: 15672
    protocol: TCP
    targetPort: http-stats
  selector:
    app.kubernetes.io/instance: rabbitmq
    app.kubernetes.io/name: rabbitmq
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```

Logs from the cis controller pod

```
2023/01/18 18:48:40 [DEBUG] Updating VirtualServer Status with { Ok} for resource name:dev2.messagebus.p2.internal , namespace: data
2023/01/18 18:48:40 [DEBUG] Enqueueing VirtualServer: &{{ } {dev2.messagebus.p2.internal  data  7ee7a089-faf6-4bed-b1aa-5e5283e5730b 220277508 1 2023-01-18 18:48:20 +0000 UTC <nil> <nil> map[f5cr:true] map[] [] []  [{kubectl-create Update cis.f5.com/v1 2023-01-18 18:48:20 +0000 UTC FieldsV1 {"f:metadata":{"f:labels":{".":{},"f:f5cr":{}}},"f:spec":{".":{},"f:pools":{},"f:tlsProfileName":{},"f:virtualServerAddress":{},"f:virtualServerName":{}}}} {k8s-bigip-ctlr.real Update cis.f5.com/v1 2023-01-18 18:48:40 +0000 UTC FieldsV1 {"f:status":{".":{},"f:status":{},"f:vsAddress":{}}}}]} {  10.40.15.207  dev2_messagebus_p2_internal_vs 0 0 [{ / rabbitmq 15672  {https /  20 31 0  } []   }] dev2.messagebus.p2.internal     [] [] []      {{ }      []  } []} {None Ok}}
2023/01/18 18:48:40 [DEBUG] Processing Key: &{data VirtualServer dev2.messagebus.p2.internal 0xc000e31180 Update}
2023/01/18 18:48:40 [DEBUG] Process all the Virtual Servers which share same VirtualServerAddress
2023/01/18 18:48:40 [DEBUG] Processing Virtual Server dev2.messagebus.p2.internal for port 443
2023/01/18 18:48:40 [DEBUG] Configured rule: {vs__rabbitmq_15672_data  0 [0xc0004b7aa0] []}
2023/01/18 18:48:40 [DEBUG] Configured policy: {dev2_messagebus_p2_internal_vs_443__policy data  [forwarding]  true [http] [0xc0004b7b00] /Common/first-match}
2023/01/18 18:48:40 [DEBUG] Processing  BIGIP referenced profiles for 'VirtualServer' 'data'/'dev2.messagebus.p2.internal'
2023/01/18 18:48:40 [DEBUG] Updated BIGIP referenced profiles for 'VirtualServer' 'data'/'dev2.messagebus.p2.internal'
2023/01/18 18:48:40 [DEBUG] Updated Virtual dev2.messagebus.p2.internal with TLSProfile dev2.messagebus.p2.internal
2023/01/18 18:48:40 [ERROR] [CORE]Endpoints could not be fetched for service rabbitmq with targetPort 0
2023/01/18 18:48:40 [DEBUG] Finished syncing virtual servers &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:dev2.messagebus.p2.internal GenerateName: Namespace:data SelfLink: UID:7ee7a089-faf6-4bed-b1aa-5e5283e5730b ResourceVersion:220277508 Generation:1 CreationTimestamp:2023-01-18 18:48:20 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[f5cr:true] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName: ManagedFields:[{Manager:kubectl-create Operation:Update APIVersion:cis.f5.com/v1 Time:2023-01-18 18:48:20 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:labels":{".":{},"f:f5cr":{}}},"f:spec":{".":{},"f:pools":{},"f:tlsProfileName":{},"f:virtualServerAddress":{},"f:virtualServerName":{}}}} {Manager:k8s-bigip-ctlr.real Operation:Update APIVersion:cis.f5.com/v1 Time:2023-01-18 18:48:40 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:status":{".":{},"f:status":{},"f:vsAddress":{}}}}]} Spec:{Host: HostGroup: VirtualServerAddress:10.40.15.207 IPAMLabel: VirtualServerName:dev2_messagebus_p2_internal_vs VirtualServerHTTPPort:0 VirtualServerHTTPSPort:0 Pools:[{Name: Path:/ Service:rabbitmq ServicePort:15672 NodeMemberLabel: Monitor:{Type:https Send:/ Recv: Interval:20 Timeout:31 TargetPort:0 Name: Reference:} Monitors:[] Rewrite: Balance: ServiceNamespace:}] TLSProfileName:dev2.messagebus.p2.internal HTTPTraffic: SNAT: WAF: RewriteAppRoot: AllowVLANs:[] IRules:[] ServiceIPAddress:[] PolicyName: PersistenceProfile: ProfileMultiplex: DOS: BotDefense: Profiles:{TCP:{Client: Server:} UDP: HTTP: HTTP2: RewriteProfile: PersistenceProfile: LogProfiles:[] ProfileL4: ProfileMultiplex:} AllowSourceRange:[]} Status:{VSAddress:None StatusOk:Ok}} (157.741µs)
```
There are no current members in the pool

![image](https://user-images.githubusercontent.com/73351998/213269581-adf2331c-44f7-4828-87d9-1616451e139d.png)


Kindly advice

@trinaths Yes, it works if I replace the name reference in targetPort to normal integer in service definition.
I'm not sure how to add port name in VS CRD; can you please share an example ?

VS
```
kg vs/dev2.messagebus.p2.internal -o yaml
apiVersion: cis.f5.com/v1
kind: VirtualServer
metadata:
  creationTimestamp: "2023-01-18T18:48:20Z"
  generation: 1
  labels:
    f5cr: "true"
  name: dev2.messagebus.p2.internal
  namespace: data
  resourceVersion: "220277508"
  uid: 7ee7a089-faf6-4bed-b1aa-5e5283e5730b
spec:
  pools:
  - monitor:
      interval: 20
      recv: ""
      send: /
      timeout: 31
      type: https
    path: /
    service: rabbitmq
    servicePort: 15672
  tlsProfileName: dev2.messagebus.p2.internal
  virtualServerAddress: 10.40.15.207
  virtualServerName: dev2_messagebus_p2_internal_vs
status:
  status: Ok
  vsAddress: None
```

service
```
 kg svc/rabbitmq -o yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    meta.helm.sh/release-name: rabbitmq
    meta.helm.sh/release-namespace: data
  creationTimestamp: "2022-09-05T20:08:06Z"
  labels:
    app.kubernetes.io/instance: rabbitmq
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: rabbitmq
    helm.sh/chart: rabbitmq-8.31.5
  name: rabbitmq
  namespace: data
  resourceVersion: "220282944"
  uid: 0bc80adc-996b-4f98-962c-8798e5e1b2a9
spec:
  clusterIP: 10.43.124.161
  clusterIPs:
  - 10.43.124.161
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: amqp
    port: 5672
    protocol: TCP
    targetPort: 5672
  - name: amqp-ssl
    port: 5671
    protocol: TCP
    targetPort: 5671
  - name: epmd
    port: 4369
    protocol: TCP
    targetPort: 4369
  - name: dist
    port: 25672
    protocol: TCP
    targetPort: 25672
  - name: http-stats
    port: 15672
    protocol: TCP
    targetPort: 15672
  selector:
    app.kubernetes.io/instance: rabbitmq
    app.kubernetes.io/name: rabbitmq
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```

logs from cis
```
2023/01/18 18:57:59 [DEBUG] [CORE] NodePoller (0xc0001d15f0) notifying listener: {l:0xc00055b800 s:0xc00055b860}
2023/01/18 18:57:59 [DEBUG] [CORE] NodePoller (0xc0001d15f0) listener callback - num items: 7 err: <nil>
2023/01/18 18:58:13 [DEBUG] [2023-01-18 18:58:13,680 __main__ DEBUG] config handler woken for reset
2023/01/18 18:58:13 [DEBUG] [2023-01-18 18:58:13,680 __main__ DEBUG] loaded configuration file successfully
2023/01/18 18:58:13 [DEBUG] [2023-01-18 18:58:13,680 __main__ DEBUG] NET Config: {}
2023/01/18 18:58:13 [DEBUG] [2023-01-18 18:58:13,680 __main__ DEBUG] loaded configuration file successfully
2023/01/18 18:58:13 [DEBUG] [2023-01-18 18:58:13,681 __main__ DEBUG] updating tasks finished, took 0.0007417201995849609 seconds
2023/01/18 18:58:26 [DEBUG] Enqueueing Old Service: &Service{ObjectMeta:{rabbitmq  data  0bc80adc-996b-4f98-962c-8798e5e1b2a9 220276321 0 2022-09-05 20:08:06 +0000 UTC <nil> <nil> map[app.kubernetes.io/instance:rabbitmq app.kubernetes.io/managed-by:Helm app.kubernetes.io/name:rabbitmq helm.sh/chart:rabbitmq-8.31.5] map[meta.helm.sh/release-name:rabbitmq meta.helm.sh/release-namespace:data] [] []  [{helm Update v1 2022-09-05 20:08:06 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{".":{},"f:meta.helm.sh/release-name":{},"f:meta.helm.sh/release-namespace":{}},"f:labels":{".":{},"f:app.kubernetes.io/instance":{},"f:app.kubernetes.io/managed-by":{},"f:app.kubernetes.io/name":{},"f:helm.sh/chart":{}}},"f:spec":{"f:internalTrafficPolicy":{},"f:ports":{".":{},"k:{\"port\":4369,\"protocol\":\"TCP\"}":{".":{},"f:name":{},"f:port":{},"f:protocol":{}},"k:{\"port\":5671,\"protocol\":\"TCP\"}":{".":{},"f:name":{},"f:port":{},"f:protocol":{}},"k:{\"port\":5672,\"protocol\":\"TCP\"}":{".":{},"f:name":{},"f:port":{},"f:protocol":{}},"k:{\"port\":15672,\"protocol\":\"TCP\"}":{".":{},"f:name":{},"f:port":{},"f:protocol":{}},"k:{\"port\":25672,\"protocol\":\"TCP\"}":{".":{},"f:name":{},"f:port":{},"f:protocol":{}}},"f:selector":{},"f:sessionAffinity":{},"f:type":{}}}} {kubectl-edit Update v1 2023-01-17 17:02:41 +0000 UTC FieldsV1 {"f:spec":{"f:ports":{"k:{\"port\":4369,\"protocol\":\"TCP\"}":{"f:targetPort":{}},"k:{\"port\":5671,\"protocol\":\"TCP\"}":{"f:targetPort":{}},"k:{\"port\":5672,\"protocol\":\"TCP\"}":{"f:targetPort":{}},"k:{\"port\":15672,\"protocol\":\"TCP\"}":{"f:targetPort":{}},"k:{\"port\":25672,\"protocol\":\"TCP\"}":{"f:targetPort":{}}}}}}]},Spec:ServiceSpec{Ports:[]ServicePort{ServicePort{Name:amqp,Protocol:TCP,Port:5672,TargetPort:{1 0 amqp},NodePort:0,AppProtocol:nil,},ServicePort{Name:amqp-ssl,Protocol:TCP,Port:5671,TargetPort:{1 0 amqp-ssl},NodePort:0,AppProtocol:nil,},ServicePort{Name:epmd,Protocol:TCP,Port:4369,TargetPort:{1 0 epmd},NodePort:0,AppProtocol:nil,},ServicePort{Name:dist,Protocol:TCP,Port:25672,TargetPort:{1 0 dist},NodePort:0,AppProtocol:nil,},ServicePort{Name:http-stats,Protocol:TCP,Port:15672,TargetPort:{1 0 http-stats},NodePort:0,AppProtocol:nil,},},Selector:map[string]string{app.kubernetes.io/instance: rabbitmq,app.kubernetes.io/name: rabbitmq,},ClusterIP:10.43.124.161,Type:ClusterIP,ExternalIPs:[],SessionAffinity:None,LoadBalancerIP:,LoadBalanc
erSourceRanges:[],ExternalName:,ExternalTrafficPolicy:,HealthCheckNodePort:0,PublishNotReadyAddresses:false,SessionAffinityConfig:nil,TopologyKeys:[],IPFamilyPolicy:*SingleStack,ClusterIPs:[10.43.124.161],IPFamilies:[IPv4],AllocateLoadBalancerNodePorts:nil,LoadBalancerClass:nil,InternalTrafficPolicy:*Cluster,},Status:ServiceStatus{LoadBalancer:LoadBalancerStatus{Ingress:[]LoadBalancerIngress{},},Conditions:[]Condition{},},}
2023/01/18 18:58:26 [DEBUG] Enqueueing Updated Service: &Service{ObjectMeta:{rabbitmq  data  0bc80adc-996b-4f98-962c-8798e5e1b2a9 220282944 0 2022-09-05 20:08:06 +0000 UTC <nil> <nil> map[app.kubernetes.io/instance:rabbitmq app.kubernetes.io/managed-by:Helm app.kubernetes.io/name:rabbitmq helm.sh/chart:rabbitmq-8.31.5] map[meta.helm.sh/release-name:rabbitmq meta.helm.sh/release-namespace:data] [] []  [{helm Update v1 2022-09-05 20:08:06 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{".":{},"f:meta.helm.sh/release-name":{},"f:meta.helm.sh/release-namespace":{}},"f:labels":{".":{},"f:app.kubernetes.io/instance":{},"f:app.kubernetes.io/managed-by":{},"f:app.kubernetes.io/name":{},"f:helm.sh/chart":{}}},"f:spec":{"f:internalTrafficPolicy":{},"f:ports":{".":{},"k:{\"port\":4369,\"protocol\":\"TCP\"}":{".":{},"f:name":{},"f:port":{},"f:protocol":{}},"k:{\"port\":5671,\"protocol\":\"TCP\"}":{".":{},"f:name":{},"f:port":{},"f:protocol":{}},"k:{\"port\":5672,\"protocol\":\"TCP\"}":{".":{},"f:name":{},"f:port":{},"f:protocol":{}},"k:{\"port\":15672,\"protocol\":\"TCP\"}":{".":{},"f:name":{},"f:port":{},"f:protocol":{}},"k:{\"port\":25672,\"protocol\":\"TCP\"}":{".":{},"f:name":{},"f:port":{},"f:protocol":{}}},"f:selector":{},"f:sessionAffinity":{},"f:type":{}}}} {kubectl-edit Update v1 2023-01-17 17:02:41 +0000 UTC FieldsV1 {"f:spec":{"f:ports":{"k:{\"port\":4369,\"protocol\":\"TCP\"}":{"f:targetPort":{}},"k:{\"port\":5671,\"protocol\":\"TCP\"}":{"f:targetPort":{}},"k:{\"port\":5672,\"protocol\":\"TCP\"}":{"f:targetPort":{}},"k:{\"port\":15672,\"protocol\":\"TCP\"}":{"f:targetPort":{}},"k:{\"port\":25672,\"protocol\":\"TCP\"}":{"f:targetPort":{}}}}}}]},Spec:ServiceSpec{Ports:[]ServicePort{ServicePort{Name:amqp,Protocol:TCP,Port:5672,TargetPort:{0 5672 },NodePort:0,AppProtocol:nil,},ServicePort{Name:amqp-ssl,Protocol:TCP,Port:5671,TargetPort:{0 5671 },NodePort:0,AppProtocol:nil,},ServicePort{Name:epmd,Protocol:TCP,Port:4369,TargetPort:{0 4369 },NodePort:0,AppProtocol:nil,},ServicePort{Name:dist,Protocol:TCP,Port:25672,TargetPort:{0 25672 },NodePort:0,AppProtocol:nil,},ServicePort{Name:http-stats,Protocol:TCP,Port:15672,TargetPort:{0 15672 },NodePort:0,AppProtocol:nil,},},Selector:map[string]string{app.kubernetes.io/instance: rabbitmq,app.kubernetes.io/name: rabbitmq,},ClusterIP:10.43.124.161,Type:ClusterIP,ExternalIPs:[],SessionAffinity:None,LoadBalancerIP:,LoadBalancerSourceRanges:[],ExternalName:,ExternalTrafficPolicy:,HealthCheckNodePort:0,PublishNotReadyAddresses:false,SessionAffinityConfig:nil,TopologyKeys:[],IPFamilyPolicy:*SingleStack,ClusterIPs:[10.43.124.161],IPFamilies:[IPv4],AllocateLoadBalancerNodePorts:nil,LoadBalancerClass:nil,InternalTrafficPolicy:*Cluster,},Status:ServiceStatus{LoadBalancer:LoadBalancerStatus{Ingress:[]LoadBalancerIngress{},},Conditions:[]Condition{},},}
2023/01/18 18:58:26 [DEBUG] Processing Key: &{data Service rabbitmq 0xc0001d8780 Delete}
2023/01/18 18:58:26 [DEBUG] Process all the Virtual Servers which share same VirtualServerAddress
2023/01/18 18:58:26 [DEBUG] Processing Virtual Server dev2.messagebus.p2.internal for port 443
2023/01/18 18:58:26 [DEBUG] Configured rule: {vs__rabbitmq_15672_data  0 [0xc0007f4780] []}
2023/01/18 18:58:26 [DEBUG] Configured policy: {dev2_messagebus_p2_internal_vs_443__policy data  [forwarding]  true [http] [0xc0007f47e0] /Common/first-match}
2023/01/18 18:58:26 [DEBUG] Processing  BIGIP referenced profiles for 'VirtualServer' 'data'/'dev2.messagebus.p2.internal'
2023/01/18 18:58:26 [DEBUG] Updated BIGIP referenced profiles for 'VirtualServer' 'data'/'dev2.messagebus.p2.internal'
2023/01/18 18:58:26 [DEBUG] Updated Virtual dev2.messagebus.p2.internal with TLSProfile dev2.messagebus.p2.internal
2023/01/18 18:58:26 [DEBUG] Finished syncing virtual servers &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:dev2.messagebus.p2.internal GenerateName: Namespace:data SelfLink: UID:7ee7a089-faf6-4bed-b1aa-5e5283e5730b ResourceVersion:220277508 Generation:1 CreationTimestamp:2023-01-18 18:48:20 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[f5cr:true] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName: ManagedFields:[{Manager:kubectl-create Operation:Update APIVersion:cis.f5.com/v1 Time:2023-01-18 18:48:20 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:labels":{".":{},"f:f5cr":{}}},"f:spec":{".":{},"f:pools":{},"f:tlsProfileName":{},"f:virtualServerAddress":{},"f:virtualServerName":{}}}} {Manager:k8s-bigip-ctlr.real Operation:Update APIVersion:cis.f5.com/v1 Time:2023-01-18 18:48:40 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:status":{".":{},"f:status":{},"f:vsAddress":{}}}}]} Spec:{Host: HostGroup: VirtualServerAddress:10.40.15.207 IPAMLabel: VirtualServerName:dev2_messagebus_p2_internal_vs VirtualServerHTTPPort:0 VirtualServerHTTPSPort:0 Pools:[{Name: Path:/ Service:rabbitmq ServicePort:15672 NodeMemberLabel: Monitor:{Type:https Send:/ Recv: Interval:20 Timeout:31 TargetPort:0 Name: Reference:} Monitors:[] Rewrite: Balance: ServiceNamespace:}] TLSProfileName:dev2.messagebus.p2.internal HTTPTraffic: SNAT: WAF: RewriteAppRoot: AllowVLANs:[] IRules:[] ServiceIPAddress:[] PolicyName: PersistenceProfile: ProfileMultiplex: DOS: BotDefense: Profiles:{TCP:{Client: Server:} UDP: HTTP: HTTP2: RewriteProfile: PersistenceProfile: LogProfiles:[] ProfileL4: ProfileMultiplex:} AllowSourceRange:[]} Status:{VSAddress:None StatusOk:Ok}} (148.434µs)
2023/01/18 18:58:26 [DEBUG] Change in Service rabbitmq does not effect any VirtualServer for TransportServer
2023/01/18 18:58:26 [DEBUG] Processing Key: &{data Service rabbitmq 0xc00077a280 Create}
2023/01/18 18:58:26 [INFO] No IngressLink found in namespace data
2023/01/18 18:58:26 [DEBUG] Process all the Virtual Servers which share same VirtualServerAddress
2023/01/18 18:58:26 [DEBUG] Processing Virtual Server dev2.messagebus.p2.internal for port 443
2023/01/18 18:58:26 [DEBUG] Configured rule: {vs__rabbitmq_15672_data  0 [0xc0007f4960] []}
2023/01/18 18:58:26 [DEBUG] Configured policy: {dev2_messagebus_p2_internal_vs_443__policy data  [forwarding]  true [http] [0xc0007f49c0] /Common/first-match}
2023/01/18 18:58:26 [DEBUG] Processing  BIGIP referenced profiles for 'VirtualServer' 'data'/'dev2.messagebus.p2.internal'
2023/01/18 18:58:26 [DEBUG] Updated BIGIP referenced profiles for 'VirtualServer' 'data'/'dev2.messagebus.p2.internal'
2023/01/18 18:58:26 [DEBUG] Updated Virtual dev2.messagebus.p2.internal with TLSProfile dev2.messagebus.p2.internal
2023/01/18 18:58:26 [ERROR] [CORE]Endpoints could not be fetched for service rabbitmq with targetPort 15672
2023/01/18 18:58:26 [INFO] No IngressLink found in namespace data
2023/01/18 18:58:26 [DEBUG] Finished syncing virtual servers &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:dev2.messagebus.p2.internal GenerateName: Namespace:data SelfLink: UID:7ee7a089-faf6-4bed-b1aa-5e5283e5730b ResourceVersion:220277508 Generation:1 CreationTimestamp:2023-01-18 18:48:20 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[f5cr:true] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName: ManagedFields:[{Manager:kubectl-create Operation:Update APIVersion:cis.f5.com/v1 Time:2023-01-18 18:48:20 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:labels":{".":{},"f:f5cr":{}}},"f:spec":{".":{},"f:pools":{},"f:tlsProfileName":{},"f:virtualServerAddress":{},"f:virtualServerName":{}}}} {Manager:k8s-bigip-ctlr.real Operation:Update APIVersion:cis.f5.com/v1 Time:2023-01-18 18:48:40 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:status":{".":{},"f:status":{},"f:vsAddress":{}}}}]} Spec:{Host: HostGroup: VirtualServerAddress:10.40.15.207 IPAMLabel: VirtualServerName:dev2_messagebus_p2_internal_vs VirtualServerHTTPPort:0 VirtualServerHTTPSPort:0 Pools:[{Name: Path:/ Service:rabbitmq ServicePort:15672 NodeMemberLabel: Monitor:{Type:https Send:/ Recv: Interval:20 Timeout:31 TargetPort:0 Name: Reference:} Monitors:[] Rewrite: Balance: ServiceNamespace:}] TLSProfileName:dev2.messagebus.p2.internal HTTPTraffic: SNAT: WAF: RewriteAppRoot: AllowVLANs:[] IRules:[] ServiceIPAddress:[] PolicyName: PersistenceProfile: ProfileMultiplex: DOS: BotDefense: Profiles:{TCP:{Client: Server:} UDP: HTTP: HTTP2: RewriteProfile: PersistenceProfile: LogProfiles:[] ProfileL4: ProfileMultiplex:} AllowSourceRange:[]} Status:{VSAddress:None StatusOk:Ok}} (105.124µs)
2023/01/18 18:58:26 [DEBUG] Change in Service rabbitmq does not effect any VirtualServer for TransportServer
2023/01/18 18:58:26 [DEBUG] [CCCL] ConfigWriter (0xc0004ff320) writing section name gtm
2023/01/18 18:58:26 [DEBUG] [CCCL] ConfigWriter (0xc0004ff320) successfully wrote section (gtm)
2023/01/18 18:58:26 [DEBUG] Wrote gtm config section: map[]
2023/01/18 18:58:26 [DEBUG] [AS3] No change in dev2-p2 tenant configuration
2023/01/18 18:58:26 [DEBUG] Enqueueing Endpoints: &Endpoints{ObjectMeta:{rabbitmq  data  8e55aa3c-cd37-4afe-8a27-1abc4136febc 220282945 0 2022-09-05 20:08:06 +0000 UTC <nil> <nil> map[app.kubernetes.io/instance:rabbitmq app.kubernetes.io/managed-by:Helm app.kubernetes.io/name:rabbitmq helm.sh/chart:rabbitmq-8.31.5] map[] [] []  [{kube-controller-manager Update v1 2022-10-28 12:09:07 +0000 UTC FieldsV1 {"f:metadata":{"f:labels":{".":{},"f:app.kubernetes.io/instance":{},"f:app.kubernetes.io/managed-by":{},"f:app.kubernetes.io/name":{},"f:helm.sh/chart":{}}},"f:subsets":{}}}]},Subsets:[]EndpointSubset{EndpointSubset{Addresses:[]EndpointAddress{EndpointAddress{IP:10.42.129.48,TargetRef:&ObjectReference{Kind:Pod,Namespace:data,Name:rabbitmq-0,UID:70d0707e-fea9-44b0-8631-06cb1e95b381,APIVersion:,ResourceVersion:165158358,FieldPath:,},Hostname:,NodeName:*d2-cont-wkr1,},EndpointAddress{IP:10.42.209.239,TargetRef:&ObjectReference{Kind:Pod,Namespace:data,Name:rabbitmq-2,UID:7d3107a8-c4d2-4c01-9622-55ea06f37390,APIVersion:,ResourceVersion:149542348,FieldPath:,},Hostname:,NodeName:*d2-cont-wkr3,},EndpointAddress{IP:10.42.92.104,TargetRef:&ObjectReference{Kind:Pod,Namespace:data,Name:rabbitmq-1,UID:00461768-68f5-4ff3-a04c-a51ec7434d80,APIVersion:,ResourceVersion:214939094,FieldPath:,},Hostname:,NodeName:*d2-cont-wkr2,},},NotReadyAddresses:[]EndpointAddress{},Ports:[]EndpointPort{EndpointPort{Name:amqp,Port:5672,Protocol:TCP,AppProtocol:nil,},EndpointPort{Name:amqp-ssl,Port:5671,Protocol:TCP,AppProtocol:nil,},EndpointPort{Name:http-stats,Port:15672,Protocol:TCP,AppProtocol:nil,},EndpointPort{Name:epmd,Port:4369,Protocol:TCP,AppProtocol:nil,},EndpointPort{Name:dist,Port:25672,Protocol:TCP,AppProtocol:nil,},},},},}
2023/01/18 18:58:26 [DEBUG] Processing Key: &{data Endpoints rabbitmq 0xc001252140 Update}
2023/01/18 18:58:26 [DEBUG] [CCCL] ConfigWriter (0xc0004ff320) writing section name gtm
2023/01/18 18:58:26 [DEBUG] [CCCL] ConfigWriter (0xc0004ff320) successfully wrote section (gtm)
2023/01/18 18:58:26 [DEBUG] Wrote gtm config section: map[]
2023/01/18 18:58:26 [DEBUG] [AS3] PostManager Accepted the configuration
2023/01/18 18:58:26 [DEBUG] [AS3] posting request to https://10.40.12.16/mgmt/shared/appsvcs/declare/dev2-p2
```

--- a/config/apis/cis/v1/types.go
+++ b/config/apis/cis/v1/types.go
@@ -2,6 +2,7 @@ package v1
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
 // +genclient
@@ -65,18 +66,18 @@ type ServiceAddress struct {
 
 // Pool defines a pool object in BIG-IP.
 type Pool struct {
-	Name              string    `json:"name,omitempty"`
-	Path              string    `json:"path,omitempty"`
-	Service           string    `json:"service"`
-	ServicePort       int32     `json:"servicePort"`
-	NodeMemberLabel   string    `json:"nodeMemberLabel,omitempty"`
-	Monitor           Monitor   `json:"monitor"`
-	Monitors          []Monitor `json:"monitors"`
-	Rewrite           string    `json:"rewrite,omitempty"`
-	Balance           string    `json:"loadBalancingMethod,omitempty"`
-	ServiceNamespace  string    `json:"serviceNamespace,omitempty"`
-	ReselectTries     int32     `json:"reselectTries,omitempty"`
-	ServiceDownAction string    `json:"serviceDownAction,omitempty"`
+	Name              string             `json:"name,omitempty"`
+	Path              string             `json:"path,omitempty"`
+	Service           string             `json:"service"`
+	ServicePort       intstr.IntOrString `json:"servicePort"`
+	NodeMemberLabel   string             `json:"nodeMemberLabel,omitempty"`
+	Monitor           Monitor            `json:"monitor"`
+	Monitors          []Monitor          `json:"monitors"`
+	Rewrite           string             `json:"rewrite,omitempty"`
+	Balance           string             `json:"loadBalancingMethod,omitempty"`
+	ServiceNamespace  string             `json:"serviceNamespace,omitempty"`
+	ReselectTries     int32              `json:"reselectTries,omitempty"`
+	ServiceDownAction string             `json:"serviceDownAction,omitempty"`
 }
 
 // Monitor defines a monitor object in BIG-IP.

--- a/docs/RELEASE-NOTES.rst
+++ b/docs/RELEASE-NOTES.rst
@@ -21,6 +21,7 @@ Added Functionality
         * `Issue 2686 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2686>`_: Validate insecure Virtual Server CR
         * `Issue 2666 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2666>`_: Support multiple virtual addresses on VirtualServer CR
         * `Issue 2703 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2703>`_: Support host group having multiple hosts with EDNS
+        * `Issue 2729 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2729>`_: Support for named port with servicePort
     * `Issue 2677 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2677>`_: Remove NotReady state nodes from BIGIP poolmembers in NodePortMode
 
 Bug Fixes

--- a/docs/config_examples/customResource/CustomResource.md
+++ b/docs/config_examples/customResource/CustomResource.md
@@ -71,7 +71,7 @@ This page is created to document the behaviour of CIS in CRD Mode.
 | path             | String  | Required | NA      | Path to access the service                                                                                                              |
 | service          | String  | Required | NA      | Service deployed in kubernetes cluster                                                                                                  |
 | nodeMemberLabel  | String  | Optional | NA      | List of Nodes to consider in NodePort Mode as BIG-IP pool members. This Option is only applicable for NodePort Mode                     |
-| servicePort      | String  | Required | NA      | Port to access Service                                                                                                                  |
+| servicePort      | Integer or String  | Required | NA      | Port to access Service.Could be service port, service port name or targetPort of the service                                 |                                                                                |
 | monitor          | monitor | Optional | NA      | Health Monitor to check the health of Pool Members                                                                                      |
 | monitors         | monitor | Optional | NA      | Specifies multiple monitors for VS Pool                                                                                                 |
 | rewrite          | String  | Optional | NA      | Rewrites the path in the HTTP Header while submitting the request to Server in the pool                                                 |
@@ -173,7 +173,7 @@ different terminations(for same domain), one with edge and another with re-encry
 | PARAMETER | TYPE    | REQUIRED | DEFAULT | DESCRIPTION                                        |
 | ------ |---------| ------ | ------ |----------------------------------------------------|
 | service | String  | Required | NA | Service deployed in kubernetes cluster             |
-| servicePort | String  | Required | NA | Port to access Service                             |
+| servicePort | Integer or String  | Required | NA | Port to access Service.Could be service port, service port name or targetPort of the service|
 | monitor | monitor  | Optional | NA | Health Monitor to check the health of Pool Members |
 | monitors | monitor | Optional | NA | Specifies multiple monitors for TS Pool            |
 | nodeMemberLabel  | String  | Optional | NA      | List of Nodes to consider in NodePort Mode as BIG-IP pool members. This Option is only applicable for NodePort Mode                     |

--- a/docs/config_examples/customResource/TransportServer/ts-with-named-port.yaml
+++ b/docs/config_examples/customResource/TransportServer/ts-with-named-port.yaml
@@ -1,0 +1,21 @@
+apiVersion: cis.f5.com/v1
+kind: TransportServer
+metadata:
+  labels:
+    f5cr: "true"
+  name: cr-transport-server
+  namespace: default
+spec:
+  mode: standard
+  pool:
+    monitor:
+      interval: 20
+      timeout: 10
+      type: udp
+    service: pytest-svc-1
+    # servicePort could be port of the service, service port name or targetPort of the serviceport
+    servicePort: port-udp # tagetPort defined on service.
+  snat: auto
+  type: udp
+  virtualServerAddress: 10.8.1.107
+  virtualServerPort: 1344

--- a/docs/config_examples/customResource/VirtualServer/virtual-with-named-port/README.md
+++ b/docs/config_examples/customResource/VirtualServer/virtual-with-named-port/README.md
@@ -1,0 +1,16 @@
+# Virtual Server with named port
+
+This section demonstrates the option to configure virtual server with servicePort with either service port name or named targetPort of service. 
+servicePort can be either port of the service listening on, service port name or targetPort defined on service. 
+
+
+Option which can be used to configure is :
+    `servicePort`
+
+## virtual-with-service-port-name.yaml
+
+By deploying this yaml file in your cluster, CIS will create a Virtual Server with endpoints matching service port name defined in virtual server.
+
+## virtual-with-target-port-name.yaml
+
+By deploying this yaml file in your cluster, CIS will create a Virtual Server with endpoints matching target port name defined in virtual server.

--- a/docs/config_examples/customResource/VirtualServer/virtual-with-named-port/sample-deployment.yaml
+++ b/docs/config_examples/customResource/VirtualServer/virtual-with-named-port/sample-deployment.yaml
@@ -1,0 +1,27 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    name: svc-1
+  name: svc-1
+  namespace: default
+spec:
+  selector:
+    matchLabels:
+      app: pytest-svc-1
+  template:
+    metadata:
+      labels:
+        app: pytest-svc-1
+    spec:
+      containers:
+        - env:
+            - name: service_name
+              value: svc-1
+          image: f5networksdevel/test-nginx:latest
+          imagePullPolicy: IfNotPresent
+          name: svc-1
+          ports:
+            - containerPort: 80
+              name: port-80
+              protocol: TCP

--- a/docs/config_examples/customResource/VirtualServer/virtual-with-named-port/sample-service.yaml
+++ b/docs/config_examples/customResource/VirtualServer/virtual-with-named-port/sample-service.yaml
@@ -1,0 +1,16 @@
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    app: svc-1
+  name: svc-1
+  namespace: default
+spec:
+  ports:
+    - name: svc-1-80
+      port: 80
+      protocol: TCP
+      targetPort: port-80
+  selector:
+    app: pytest-svc-1
+  type: ClusterIP

--- a/docs/config_examples/customResource/VirtualServer/virtual-with-named-port/virtual-with-service-port-name.yaml
+++ b/docs/config_examples/customResource/VirtualServer/virtual-with-named-port/virtual-with-service-port-name.yaml
@@ -1,0 +1,16 @@
+apiVersion: "cis.f5.com/v1"
+kind: VirtualServer
+metadata:
+  name: my-new-virtual-server
+  labels:
+    f5cr: "true"
+spec:
+  # This is an insecure virtual, Please use TLSProfile to secure the virtual
+  # check out tls examples to understand more.
+  host: caffee.example.com
+  virtualServerAddress: "172.16.3.8"
+  pools:
+  - path: /coffee
+    service: svc-1
+    # servicePort could be port of the service, service port name or targetPort of the serviceport
+    servicePort: svc-1-80 # service port name

--- a/docs/config_examples/customResource/VirtualServer/virtual-with-named-port/virtual-with-target-port-name.yaml
+++ b/docs/config_examples/customResource/VirtualServer/virtual-with-named-port/virtual-with-target-port-name.yaml
@@ -1,0 +1,16 @@
+apiVersion: "cis.f5.com/v1"
+kind: VirtualServer
+metadata:
+  name: my-new-virtual-server
+  labels:
+    f5cr: "true"
+spec:
+  # This is an insecure virtual, Please use TLSProfile to secure the virtual
+  # check out tls examples to understand more.
+  host: caffee.example.com
+  virtualServerAddress: "172.16.3.8"
+  pools:
+    - path: /coffee
+      service: svc-1
+      # servicePort could be port of the service, service port name or targetPort of the serviceport
+      servicePort: port-80 # targetport of service

--- a/docs/config_examples/customResourceDefinitions/customresourcedefinitions.yml
+++ b/docs/config_examples/customResourceDefinitions/customresourcedefinitions.yml
@@ -133,9 +133,10 @@ spec:
                         type: string
                         pattern: '^[a-zA-Z0-9][-A-Za-z0-9_.\/]{0,61}[a-zA-Z0-9]=[a-zA-Z0-9][-A-Za-z0-9_.]{0,61}[a-zA-Z0-9]$'
                       servicePort:
-                        type: integer
-                        minimum: 1
-                        maximum: 65535
+                        x-kubernetes-int-or-string: true
+                        anyOf:
+                          - type: integer
+                          - type: string
                       rewrite:
                         type: string
                         pattern: '^\/([A-z0-9-_+]+\/)*([-A-z0-9_.:]+\/?)*$'
@@ -436,9 +437,10 @@ spec:
                       type: string
                       pattern: '^[a-zA-Z]+([-A-z0-9_.+])*([A-z0-9])+$'
                     servicePort:
-                      type: integer
-                      minimum: 1
-                      maximum: 65535
+                      x-kubernetes-int-or-string: true
+                      anyOf:
+                        - type: integer
+                        - type: string
                     serviceNamespace:
                       type: string
                       pattern: '^[a-zA-Z]+([-A-z0-9_.+:])*([A-z0-9])+$'

--- a/pkg/controller/resourceConfig.go
+++ b/pkg/controller/resourceConfig.go
@@ -275,7 +275,7 @@ func (ctlr *Controller) framePoolName(ns string, pool cisapiv1.Pool, host string
 
 	poolName := pool.Name
 	if poolName == "" {
-		targetPort := intstr.IntOrString{IntVal: pool.ServicePort}
+		targetPort := pool.ServicePort
 
 		if (intstr.IntOrString{}) == targetPort {
 			svcNamespace := ns
@@ -305,7 +305,7 @@ func formatPoolName(namespace, svc string, port intstr.IntOrString, nodeMemberLa
 }
 
 // format the monitor name for an VirtualServer pool
-func formatMonitorName(namespace, svc string, monitorType string, port int32, hostName string, path string) string {
+func formatMonitorName(namespace, svc string, monitorType string, port intstr.IntOrString, hostName string, path string) string {
 	monitorName := fmt.Sprintf("%s_%s", svc, namespace)
 
 	if len(hostName) > 0 {
@@ -319,8 +319,8 @@ func formatMonitorName(namespace, svc string, monitorType string, port int32, ho
 		}
 	}
 
-	if monitorType != "" && port != 0 {
-		servicePort := fmt.Sprint(port)
+	if monitorType != "" && (port.IntVal != 0 || port.StrVal != "") {
+		servicePort := fetchPortString(port)
 		monitorName = monitorName + fmt.Sprintf("_%s_%s", monitorType, servicePort)
 	}
 	return AS3NameFormatter(monitorName)
@@ -374,7 +374,7 @@ func (ctlr *Controller) deleteSvcDepResource(rsName string, rsCfg *ResourceConfi
 }
 
 // fetch target port from service
-func (ctlr *Controller) fetchTargetPort(namespace, svcName string, servicePort int32) intstr.IntOrString {
+func (ctlr *Controller) fetchTargetPort(namespace, svcName string, servicePort intstr.IntOrString) intstr.IntOrString {
 	var targetPort intstr.IntOrString
 	var svcIndexer cache.Indexer
 	svcKey := namespace + "/" + svcName
@@ -394,7 +394,11 @@ func (ctlr *Controller) fetchTargetPort(namespace, svcName string, servicePort i
 	}
 	svc := item.(*v1.Service)
 	for _, port := range svc.Spec.Ports {
-		if port.Port == servicePort {
+		if port.Port == servicePort.IntVal || port.Name == servicePort.StrVal || port.TargetPort.StrVal == servicePort.StrVal {
+			// In case of named targetPort, send service port name to match endpoint name as targetPort
+			if port.TargetPort.StrVal != "" {
+				return intstr.IntOrString{StrVal: port.Name}
+			}
 			return port.TargetPort
 		}
 	}
@@ -435,9 +439,8 @@ func (ctlr *Controller) prepareRSConfigFromVirtualServer(
 		}
 		framedPools[poolName] = struct{}{}
 		targetPort := ctlr.fetchTargetPort(vs.Namespace, pl.Service, pl.ServicePort)
-
 		if (intstr.IntOrString{}) == targetPort {
-			targetPort = intstr.IntOrString{IntVal: pl.ServicePort}
+			targetPort = pl.ServicePort
 		}
 		svcNamespace := vs.Namespace
 		if pl.ServiceNamespace != "" {
@@ -477,9 +480,9 @@ func (ctlr *Controller) prepareRSConfigFromVirtualServer(
 				if monitor.Name != "" && monitor.Reference == BIGIP {
 					pool.MonitorNames = append(pool.MonitorNames, MonitorName{Name: monitor.Name, Reference: monitor.Reference})
 				} else {
-					var formatPort int32
+					var formatPort intstr.IntOrString
 					if monitor.TargetPort != 0 {
-						formatPort = monitor.TargetPort
+						formatPort = intstr.IntOrString{IntVal: monitor.TargetPort}
 					} else {
 						formatPort = pl.ServicePort
 					}
@@ -1595,7 +1598,7 @@ func (ctlr *Controller) prepareRSConfigFromTransportServer(
 	}
 	targetPort := ctlr.fetchTargetPort(svcNamespace, vs.Spec.Pool.Service, vs.Spec.Pool.ServicePort)
 	if (intstr.IntOrString{}) == targetPort {
-		targetPort = intstr.IntOrString{IntVal: vs.Spec.Pool.ServicePort}
+		targetPort = vs.Spec.Pool.ServicePort
 	}
 
 	pool := Pool{
@@ -1634,9 +1637,9 @@ func (ctlr *Controller) prepareRSConfigFromTransportServer(
 			if monitor.Name != "" && monitor.Reference == BIGIP {
 				pool.MonitorNames = append(pool.MonitorNames, MonitorName{Name: monitor.Name, Reference: monitor.Reference})
 			} else {
-				var formatPort int32
+				var formatPort intstr.IntOrString
 				if monitor.TargetPort != 0 {
-					formatPort = monitor.TargetPort
+					formatPort = intstr.IntOrString{IntVal: monitor.TargetPort}
 				} else {
 					formatPort = pl.ServicePort
 				}
@@ -1744,9 +1747,9 @@ func (ctlr *Controller) prepareRSConfigFromLBService(
 			log.Errorf("[CORE] %s", msg)
 		}
 		pool.MonitorNames = append(pool.MonitorNames, MonitorName{Name: JoinBigipPath(rsCfg.Virtual.Partition,
-			formatMonitorName(svc.Namespace, svc.Name, monitorType, svcPort.TargetPort.IntVal, "", ""))})
+			formatMonitorName(svc.Namespace, svc.Name, monitorType, svcPort.TargetPort, "", ""))})
 		monitor = Monitor{
-			Name:      formatMonitorName(svc.Namespace, svc.Name, monitorType, svcPort.TargetPort.IntVal, "", ""),
+			Name:      formatMonitorName(svc.Namespace, svc.Name, monitorType, svcPort.TargetPort, "", ""),
 			Partition: rsCfg.Virtual.Partition,
 			Type:      monitorType,
 			Interval:  mon.Interval,

--- a/pkg/controller/resourceConfig_test.go
+++ b/pkg/controller/resourceConfig_test.go
@@ -105,7 +105,7 @@ var _ = Describe("Resource Config Tests", func() {
 			Expect(name).To(Equal("svc1_80_default_foo_app_test"), "Invalid Pool Name")
 		})
 		It("Monitor Name", func() {
-			name := formatMonitorName(namespace, "svc1", "http", 80, "foo.com", "path")
+			name := formatMonitorName(namespace, "svc1", "http", intstr.IntOrString{IntVal: 80}, "foo.com", "path")
 			Expect(name).To(Equal("svc1_default_foo_com_path_http_80"), "Invalid Monitor Name")
 		})
 		It("Rule Name", func() {
@@ -326,7 +326,7 @@ var _ = Describe("Resource Config Tests", func() {
 				cisapiv1.TransportServerSpec{
 					Pool: cisapiv1.Pool{
 						Service:          "svc1",
-						ServicePort:      80,
+						ServicePort:      intstr.IntOrString{IntVal: 80},
 						ServiceNamespace: "test",
 						Monitor: cisapiv1.Monitor{
 							Type:     "tcp",
@@ -348,7 +348,7 @@ var _ = Describe("Resource Config Tests", func() {
 				cisapiv1.TransportServerSpec{
 					Pool: cisapiv1.Pool{
 						Service:     "svc1",
-						ServicePort: 80,
+						ServicePort: intstr.IntOrString{IntVal: 80},
 						Monitors: []cisapiv1.Monitor{
 							{
 								Type:       "tcp",

--- a/pkg/controller/worker.go
+++ b/pkg/controller/worker.go
@@ -1862,7 +1862,7 @@ func (ctlr *Controller) updatePoolMembersForNPL(
 		pods := ctlr.GetPodsForService(namespace, svcName, true)
 		if pods != nil {
 			for _, svcPort := range poolMemInfo.portSpec {
-				if svcPort.TargetPort == pool.ServicePort {
+				if svcPort.TargetPort == pool.ServicePort || svcPort.Name == pool.ServicePort.StrVal {
 					podPort := svcPort.TargetPort
 					rsCfg.MetaData.Active = true
 					rsCfg.Pools[index].Members =

--- a/pkg/controller/worker_test.go
+++ b/pkg/controller/worker_test.go
@@ -1569,7 +1569,7 @@ var _ = Describe("Worker Tests", func() {
 								},
 								Rewrite:     "/bar",
 								Balance:     "fastest-node",
-								ServicePort: 80,
+								ServicePort: intstr.IntOrString{IntVal: 80},
 							},
 							{
 								Path:    "/",
@@ -1582,7 +1582,7 @@ var _ = Describe("Worker Tests", func() {
 								},
 								Rewrite:     "/bar1",
 								Balance:     "fastest-node",
-								ServicePort: 81,
+								ServicePort: intstr.IntOrString{IntVal: 81},
 							},
 						},
 						RewriteAppRoot:     "/home",
@@ -1603,7 +1603,7 @@ var _ = Describe("Worker Tests", func() {
 					},
 				)
 
-				fooPorts = []v1.ServicePort{{Port: 80, NodePort: 30001},
+				fooPorts = []v1.ServicePort{{Port: 80, NodePort: 30001, TargetPort: intstr.IntOrString{StrVal: "port-80"}},
 					{Port: 8080, NodePort: 38001},
 					{Port: 9090, NodePort: 39001}}
 				fooIps := []string{"10.1.1.1"}
@@ -1689,13 +1689,12 @@ var _ = Describe("Worker Tests", func() {
 				mockCtlr.processResources()
 				// Should process VS now
 				Expect(len(mockCtlr.resources.ltmConfig)).To(Equal(1), "Virtual Server not Processed")
-
 				mockCtlr.deleteService(svc)
 				mockCtlr.processResources()
 
 				// Update service
 				svc.Spec.Ports[0].NodePort = 30002
-				Expect(fetchPortString(intstr.IntOrString{StrVal: strconv.Itoa(int(vs.Spec.Pools[0].ServicePort))})).To(BeEquivalentTo("80"))
+				Expect(fetchPortString(intstr.IntOrString{StrVal: strconv.Itoa(int(vs.Spec.Pools[0].ServicePort.IntVal))})).To(BeEquivalentTo("80"))
 				mockCtlr.addService(svc)
 				mockCtlr.processResources()
 
@@ -1982,7 +1981,7 @@ var _ = Describe("Worker Tests", func() {
 
 			BeforeEach(func() {
 				//Add Virtual Server
-				fooPorts := []v1.ServicePort{{Port: 80, NodePort: 30001},
+				fooPorts := []v1.ServicePort{{Port: 80, NodePort: 30001, TargetPort: intstr.IntOrString{StrVal: "port-80"}},
 					{Port: 8080, NodePort: 38001},
 					{Port: 9090, NodePort: 39001}}
 				fooIps := []string{"10.1.1.1"}
@@ -2000,7 +1999,7 @@ var _ = Describe("Worker Tests", func() {
 						PolicyName:           "policy",
 						Pool: cisapiv1.Pool{
 							Service:     "svc1",
-							ServicePort: 80,
+							ServicePort: intstr.IntOrString{StrVal: "port-80"},
 							Monitor: cisapiv1.Monitor{
 								Type:     "tcp",
 								Timeout:  10,

